File: src/core/graph/cache.ts
import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export class FileHashCache {
	private cache: Map<string, string>;
	private cacheDir: string;
	private cacheFile: string;

	constructor(cacheDir: string) {
		this.cacheDir = cacheDir;
		this.cacheFile = join(cacheDir, "hashes.json");
		this.cache = new Map();
		this.load();
	}

	hasChanged(filePath: string, content: string): boolean {
		const hash = this.hash(content);
		const existing = this.cache.get(filePath);
		return existing !== hash;
	}

	update(filePath: string, content: string): void {
		this.cache.set(filePath, this.hash(content));
	}

	remove(filePath: string): void {
		this.cache.delete(filePath);
	}

	save(): void {
		mkdirSync(this.cacheDir, { recursive: true });
		const obj = Object.fromEntries(this.cache);
		writeFileSync(this.cacheFile, JSON.stringify(obj));
	}

	private load(): void {
		if (!existsSync(this.cacheFile)) return;
		const raw = readFileSync(this.cacheFile, "utf-8");
		const obj = JSON.parse(raw) as Record<string, string>;
		this.cache = new Map(Object.entries(obj));
	}

	private hash(content: string): string {
		return createHash("sha256").update(content).digest("hex");
	}
}

File: crates/graphmind-db/src/cache.rs
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

pub struct FileHashCache {
    cache: HashMap<String, String>,
    cache_dir: PathBuf,
    cache_file: PathBuf,
}

impl FileHashCache {
    pub fn new(cache_dir: &Path) -> Self {
        let cache_file = cache_dir.join("hashes.json");
        let mut fhc = Self {
            cache: HashMap::new(),
            cache_dir: cache_dir.to_path_buf(),
            cache_file,
        };
        fhc.load();
        fhc
    }

    pub fn has_changed(&self, file_path: &str, content: &str) -> bool {
        let hash = Self::hash(content);
        self.cache.get(file_path) != Some(&hash)
    }

    pub fn update(&mut self, file_path: &str, content: &str) {
        self.cache.insert(file_path.to_string(), Self::hash(content));
    }

    pub fn remove(&mut self, file_path: &str) {
        self.cache.remove(file_path);
    }

    pub fn known_files(&self) -> Vec<String> {
        self.cache.keys().cloned().collect()
    }

    pub fn save(&self) {
        fs::create_dir_all(&self.cache_dir).ok();
        let json = serde_json::to_string(&self.cache).unwrap_or_default();
        fs::write(&self.cache_file, json).ok();
    }

    fn load(&mut self) {
        if !self.cache_file.exists() {
            return;
        }
        if let Ok(raw) = fs::read_to_string(&self.cache_file) {
            if let Ok(map) = serde_json::from_str::<HashMap<String, String>>(&raw) {
                self.cache = map;
            }
        }
    }

    fn hash(content: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(content.as_bytes());
        format!("{:x}", hasher.finalize())
    }
}
